Simplify conversation DSL

raulraja opened the pull request:
Simplifies the conversation DSL avoiding lazy functions and monadic composition.
We are already in languages where statements are what defines control flow and we should favor the lang idiomatic choice.

javipacheco reviewed (changes requested):
Good job! Just a small comment

The new way to add contexts now is easier to understand, but now that we don't have blocks for contexts, should we have a new method to remove contexts from the vector store? I'm not sure, but maybe during the conversation, the user doesn't need some previous contexts added. We should think about which context we want to add in each call to the AI.

raulraja commented on kotlin/src/commonMain/kotlin/com/xebia/functional/xef/auto/DSLExtensions.kt:
```diff
@@ -1,4 +1,19 @@
+ * @param block The block of code representing the conversation logic.
+ * @return The result of the conversation execution.
+ */
+suspend inline fun <A> conversation(
```
The issue is that many modules lack proper `spotless` config. We need to review it and see if this is something we can apply globally to all modules.

raulraja commented on openai/src/commonMain/kotlin/com/xebia/functional/xef/auto/llm/openai/MockAIClient.kt:
```diff
@@ -1,131 +0,0 @@
- * Run the [AI] value to produce _either_ an [AIError], or [A]. This method uses the [mockAI] to
- * compute the different responses.
- */
-suspend fun <A> AI<A>.mock(mockAI: MockOpenAIClient): Either<AIError, A> =
```
I don't see the value of this at the moment, and I removed it because it was not in use anywhere.
We should have a similar test suite for conversations like the one you did for message token count.

raulraja commented:
@javipacheco addressed all your comments.

Regarding:

> Good job! Just minor comments
> 
> The new way to add contexts now is easier to understand, but now that we don't have blocks for contexts, should we have a new method to remove contexts from the vector store? I'm not sure, but maybe during the conversation, the user doesn't need some previous contexts added. We should consider which context we want to add in each call to the AI.

Yes, we need ways to add, search, and clear in both context and memories.
In practice, only memories are important because the context also appears in the memories.
I want to split memories out of the VectorStore as a separate interface and make the Conversation implement both.
Also, I think the store should be more generic and allow the user to set and unset values more geared to be regular storage based on SQL with in-memory, locally persistent, and remote persistent implementations all based on the same SQL or similar, if possible.

javipacheco reviewed (approved):
🚀  LGTM!

Montagon reviewed (approved):
🚀
